sw/device/lib/dif/dif_i2c.h

```diff
@@ -0,0 +1,390 @@
+ * @param scl The value to drive SCL to.
+ * @param sda The value to drive SDA to.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_override_drive_pins(const dif_i2c_t *i2c, bool scl,
+                                             bool sda);
+
+/**
+ * Returns oversampling of the last 16 values of the SCL and SDA pins, with the
+ * zeroth bit being the most recent.
+ *
+ * @param i2c An I2C device.
+ * @param scl_samples Out-param for SCL sample bits; may be null.
+ * @param sda_samples Out-param for SDA sample bits; may be null.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_override_sample_pins(const dif_i2c_t *i2c,
+                                              uint16_t *scl_samples,
+                                              uint16_t *sda_samples);
+
+/**
+ * Returns the current levels, i.e., number of entries, in the FMT and RX FIFOs.
+ * These values represent the number of entries pending for send by hardware,
+ * and entries pending for read by software, respectively.
+ *
+ * @param i2c An I2C device.
+ * @param fmt_fifo_level Out-param for the number of unsent FMT bytes; may be
+ *        null.
+ * @param rx_fifo_level Out-param for the number of unread RX bytes; may be
+ *        null.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_get_fifo_levels(const dif_i2c_t *i2c,
+                                         uint8_t *fmt_fifo_level,
+                                         uint8_t *rx_fifo_level);
+
+/**
+ * Pops an entry (a byte) off of the RX FIFO. Passing in `NULL` to the out-param
+ * will still trigger a byte pop.
+ *
+ * @param i2c An I2C device.
+ * @param byte Out-param for the popped byte; may be null.
+ * @return The result of the opeartion.
+ */
+dif_i2c_result_t dif_i2c_read_byte(const dif_i2c_t *i2c, uint8_t *byte);
+
+/**
+ * Flags for a formatted I2C byte, used by the `dif_i2c_write_byte_raw()`
+ * function.
+ */
+typedef struct dif_i2c_fmt_flags {
+  /**
+   * Causes a start signal to be sent before the byte.
+   *
+   * If a start has been issued during the current transaction, this will issue
+   * a repeated start.
+   */
+  bool start;
+  /**
+   * Causes a stop signal to be sent after the byte.
+   *
+   * This flag cannot be set along with `read` or `read_cont`.
+   */
+  bool stop;
+  /**
+   * Causes the byte to be interpreted as an unsigned number of bytes to read
+   * from the target; 0 is interpreted as 256.
+   */
+  bool read;
+  /**
+   * Requires `read` to be set; if so, once the final byte associated with this
+   * read is received, it will be acknowledged, allowing the read operation to
+   * continue.
+   */
+  bool read_cont;
+  /**
+   * By default, the hardware expects an ACK after every byte sent, and raises
+   * an exception (surfaced as the `kDifi2cIrqTypeNak` interrupt). This flag
+   * disables that behavior.
+   *
+   * This flag cannot be set along with `read` or `read_cont`.
+   */
+  bool nak_ok;
```

Comment:
nak_ok doesn't suggest that it disables an interrupt, it feels like this should have a
name more suggestive like suppress_nak_irq ?

Reply:
Hi @gkelly, are you suggesting that we change this flag name in the IP docs?

Reply:
No, I don't think this needs to change in the IP docs, but the behavior we expose in software needs to be explicit. `nak_ok` doesn't suggest that it has anything to do with interrupts, so we should take care of that at this layer.